packages/foundry/contracts/prediction-market/PredictionMarketHook.sol

```diff
@@ -0,0 +1,488 @@
+    uint256 public constant FEE = 10000; // 1%
+
+    /// @notice waiting period after a swap occurs in a pool and when a market can be settled
+    uint256 public constant SETTLEMENT_WAITING_PERIOD = 10;
+
+    /**
+     * @notice  The factory allowed to create pools with the PredictionMarketHook
+     * @dev Hook registration will fail if the pool is not created from the factory
+     */ 
+    address public immutable factory;
+    
+    /**
+     * @notice Mapping between prediction market id and the corresponding markets
+     * @dev mapping(marketId => PredictionMarket)
+     * marketId = keccak256(abi.encodePacked(pool, [sorted token0], [sorted token1], closedAtTimestamp));
+     */
+    mapping(bytes32 => PredictionMarket) public markets;
+
+    /**
+     * @notice Mapping of user positions to corresponding markets
+     * @dev mapping(marketId => mapping(userAddress => Position))
+     */
+    mapping(bytes32 => mapping(address => Position)) public positions;
+
+    /// @notice Lookup of pools registered with the hook
+    mapping(address => bool) private _isPoolRegistered;
+    
+    /**
+     * @notice Lookup of the last swap block by pool
+     * @dev Market settlement will fail if the last pool swap was done within the configured SETTLEMENT_WAITING_PERIOD.
+     * This is to mitigate price manipulation possibilities by allowing arbitrage to bring the pool back into balance
+     * if it has been brought out of balance by a malicious actor
+     */ 
+    mapping(address => uint256) private _lastActivityBlockByPool;
+
+    /**
+     * @notice Lookup of user address and pool prediction market participation end time
+     * @dev Market participants are incentivized with 0 swap fees. This lookup is used to determine when 
+     * if a user is a participant in a market priced by a given pool
+     *
+     * mapping(userAddress => mapping(pool => marketCloseTimestamp))
+     */
+    mapping(address => mapping(address => uint256)) private _predictionMarketPartipants;
+    
+
+    /**
+     * @notice A new `PredictionMarketHook` contract has been registered successfully for a given factory and pool.
+     * @dev If the registration fails the call will revert, so there will be no event.
+     * @param hooksContract This contract
+     * @param pool The pool on which the hook was registered
+     */
+    event PredictionMarketHookRegistered(address indexed hooksContract, address indexed pool);
+
+    /// @notice The pool does not support adding liquidity through donation.
+    error PoolDoesNotSupportDonation();
+
+    /// @notice The pool is not registered with the hook
+    error PoolNotFound();
+
+    /// @notice The pool token pair does not exist
+    error InvalidTokenPair();
+
+    /// @notice Time provided timestamp is in the past
+    error TimestampIsInPast();
+
+
+    constructor(
+        IVault vault,
+        address allowedFactory
+    ) VaultGuard(vault) Ownable(msg.sender) {
```

Comment:
It's Ownable, but there aren't any permissioned functions. I can think of some: maintaining a set of allowed factories, in case new WeightedPools or other types are developed that would work here.

Also, what if there's an issue with the pool, and it's paused or put into recovery mode? Perhaps you could have some kind of emergency stop for a market, that would halt a market at current prices and make everything claimable so that people could withdraw. To prevent manipulation by the owner, I imagine this would be permissionless, but would only work if the underlying pool were paused or in recovery.

Reply:
👍 on what can be Ownable

Good point, I didn't consider that scenario (paused pools / recovery mode) during development. The ability to pause markets in those cases should absolutely be in there.